src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,130 @@
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == elementData.length) {
+            elementData = buildUp();
+        }
+        elementData[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        checkIndex(index);
+        if (size == elementData.length) {
+            elementData = buildUp();
+        }
+        System.arraycopy(elementData, index, elementData, index + 1, size - index);
+        elementData[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        checkList(list);
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
+    @SuppressWarnings("unchecked")
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        checkIndexEqualsSize(index);
+        return (T) elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        checkIndexEqualsSize(index);
+        elementData[index] = value;
     }
 
     @Override
+    @SuppressWarnings("unchecked")
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        checkIndexEqualsSize(index);
+        final T oldValue = (T) elementData[index];
+        if (size - 1 > index) {
+            System.arraycopy(elementData, index + 1, elementData, index, size - index - 1);
+        }
+        elementData[size - 1] = null;
+        size--;
+        return oldValue;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int index = getElement(element);
+        if (index == -1) {
+            throw new NoSuchElementException("This element does not exist: " + element);
+        }
+        remove(index);
+        return element;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Object[] buildUp() {
+        int oldCapacity = elementData.length;
+        int newCapacity = (int) (oldCapacity * SIZE_INCREASE_FACTOR);
+        Object[] newElementData = new Object[newCapacity];
+        System.arraycopy(elementData, 0, newElementData, 0, oldCapacity);
+        return newElementData;
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Index out of range: " + index);
+        }
+    }
+
+    private void checkIndexEqualsSize(int index) {
+        if (index == size) {
+            throw new ArrayListIndexOutOfBoundsException("Index out of range: " + index);
+        }
+    }
+
+    private void checkList(List<T> list) {
+        if (list == null) {
+            throw new NullPointerException("List : " + list + " can't be null");
+        }
+    }
```

Comment:
Why do we need this check?

Reply:
If the specified collection is null

Reply:
I dont think we need this check for this task, and even if we provide it, we should throw some RuntineException, not NPE